creator-node/src/middlewares/premiumContent/premiumContentMiddleware.ts

```diff
@@ -31,70 +29,71 @@ export const premiumContentMiddleware = async (
+    const libs = req.app.get('audiusLibs')
+    const redis = req.app.get('redisClient')
     const logger = (req as any).logger as Logger
 
-    const discoveryNodeWallet = recoverWallet(
-      signedDataFromDiscoveryNode,
-      signatureFromDiscoveryNode
-    )
-    const isRegisteredDN = await isRegisteredDiscoveryNode({
-      wallet: discoveryNodeWallet,
-      libs: req.app.get('audiusLibs'),
-      logger,
-      redis: req.app.get('redisClient')
-    })
-    if (!isRegisteredDN) {
-      return sendResponse(
-        req,
-        res,
-        errorResponseForbidden(
-          'Failed discovery node signature validation for premium content.'
-        )
-      )
+    const { doesUserHaveAccess, trackId, isPremium, error } =
+      await premiumContentAccessChecker.checkPremiumContentAccess({
+        cid,
+        premiumContentHeaders,
+        libs,
+        logger,
+        redis
+      })
+    if (doesUserHaveAccess) {
+      // Set premium content track id and 'premium-ness' so that next middleware or
+      // request handler does not need to make trips to the database to get this info.
+      // We need the info because if the content is premium, then we need to set
+      // the cache-control response header to no-cache so that nginx does not cache it.
+      ;(req as any).premiumContent = {
+        trackId,
+        isPremium
+      }
+      next()
+      return
     }
 
-    const { premiumContentId, premiumContentType } = req.params
-
-    const userWallet = recoverWallet(signedDataFromUser, signatureFromUser)
-    const signedDataFromDiscoveryNodeObj = JSON.parse(
-      signedDataFromDiscoveryNode as string
-    )
-    const isMatch = isPremiumContentMatch({
-      signedDataFromDiscoveryNode: signedDataFromDiscoveryNodeObj,
-      userWallet,
-      premiumContentId: parseInt(premiumContentId),
-      premiumContentType: premiumContentType as PremiumContentType,
-      logger
-    })
-    if (!isMatch) {
-      return sendResponse(
-        req,
-        res,
-        errorResponseForbidden('Failed match verification for premium content.')
-      )
+    switch (error) {
+      case PremiumContentAccessError.MISSING_HEADERS:
+        return sendResponse(
+          req,
+          res,
+          errorResponseUnauthorized(
+            'Missing request headers for premium content.'
+          )
+        )
+      case PremiumContentAccessError.INVALID_DISCOVERY_NODE:
+        return sendResponse(
+          req,
+          res,
+          errorResponseForbidden(
+            'Failed discovery node signature validation for premium content.'
+          )
+        )
+      case PremiumContentAccessError.FAILED_MATCH:
+      default:
+        return sendResponse(
+          req,
+          res,
+          errorResponseForbidden(
+            'Failed match verification for premium content.'
+          )
+        )
     }
-
-    next()
   } catch (e) {
     const error = `Could not validate premium content access: ${
```

Comment:
Thinking through this – do we definitely want to return an error in the response here? If something fails in this block of code, every CID fetch, even for non-premium content, will fail. Would it be better to just log an error and attach metadata to the request indicating that the user doesn't have access to the premium content?

Reply:
makes sense to let through on uncaught error

Reply:
@SidSethi are you saying we should just allow access to the CID even if it's premium, if for some reason there is an uncaught error?

Reply:
how does CN know its premium if there's an uncaught error in validation? basically how does CN know if CID is premium or not - if it can't verify due to some bug in our code, then maybe we should not reject? at least at the beginning - maybe worth doing a pass to minimize surface area of uncaught errors

Reply:
i think the automated unit tests and manual testing that I've done give a pretty high confidence that there is no uncaught errors. However, there may still be, or uncaught errors may be introduced in the future. I agree with that. I think we just need to make a call if we want to proceed with giving/preventing access for the user in this (edge) case